texastribune/files
# Replace accents in filenames

#### What's this PR do?

Removes the accents from uploaded files, converting them to their corresponding non-accented ASCII characters.

Eg. "tést_ötè-12î.png" becomes "test_ote-12i.png"

#### Why are we doing this? How does it help us?

The photo dept sometimes needs to upload files that have accents in their filenames, which throws a name validation error. The photo dept therefore has to manually change these names. This PR solves that problem.

#### How should this be manually tested?

**Testing files app:**
1) In the project repo, run `make start`
2) A browser tab should open with the files app.
3) On your machine, either rename a file on your harddrive to include some accents or create a file that includes some accents in. eg `ã tęšt fíléñâmē`
4) Drag and drop the file into the files app. Ensure that the name is properly converted. It should now read `a test filename`
5) Delete the file.
6) This time, click on the cog in the file app and select `add file`. Select the same file and ensure that it is again uploaded successfully with the converted name.
7) Add a few files with ordinary, non-accented file names, to ensure that those upload normally.

**Testing in Tribune app:**
Because the Tribune app makes some complicated modifications to the behavior of the files app, it's important to ensure that its working correctly when imported there. To test:

1) In the Tribune project, create a new branch from master: `git checkout -b test-files-app`
2) Spin up the docker container.
3) When the container has spun up and the docker shell has opened. You should be in the `app` directory. Use the terminal to navigate up one directory: `cd ..`
4) Enter the `node_modules` directory: `cd node_modules`.
5) In that directory, install this files app branch: `npm i ttfiles@https://github.com/texastribune/files.git#websafe-filenames`
6) Once installed, navigate back to the the `app` directory and run `npm run start`
7) Navigate to the files app in your browser: http://local.texastribune.org:8000/admin/tt_files/mediafile/
8) Repeat the tests above from "Testing files app" to ensure file name conversion works successfully.

#### How should this change be communicated to end users?

We should tell the photo team.

#### Are there any smells or added technical debt to note?

- My accent conversion func `createWebSafeString` is being called from custom elements in the `ui` directory (`src/ui/browser.ts` and `src/ui/contextMenu.ts`). I'm not sure if this is entirely preferable. There is probably a strong argument to be made that this kind of business logic should be handled in the `Directory` class or its subclasses (eg. MemoryDirectory, LocalStorageDirectory, etc), but based on the way these classes are structured it seemed like it would require fewer code changes to simply handle the file name conversion in `src/ui/browser.ts` and `src/ui/contextMenu.ts` before the name is passed on to the `addFile` method of these directory subclasses.
- Although I have called the func `createWebSafeString` I should note that this func is only replacing accents. It's not handling spaces or other special characters. I could expand the functionality of this func or I could rename it to be a bit more explicit. I wasn't sure it mattered too much so I left it as is. I'm happy to change this though if we'd like.
- As always with the Files app, when reviewing this code you can largely ignore the file changes in `lib` (the compiled JS files) and bundle.js (the minified code). The relevant changes are all in files in `src`.

#### What are the relevant tickets?

https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwKqLrNFM17Lq15R/rechCDx9r7wA8pHOu?blocks=hide

#### TODOs / next steps:
